p2p/src/peer_manager/mod.rs

```diff
@@ -123,20 +133,99 @@ where
+    /// This won't work for majority of nodes but that should be accepted.
+    async fn handle_outbound_receiver_address(
+        &mut self,
+        peer_id: T::PeerId,
+        receiver_address: PeerAddress,
+    ) -> crate::Result<()> {
+        if !self.is_peer_address_valid(&receiver_address) {
+            return Ok(());
+        }
+
+        // Take IP and use port numbers from all listening sockets (with same IP version)
+        let discovered_own_addresses = self
+            .peer_connectivity_handle
+            .local_addresses()
+            .iter()
+            .map(TransportAddress::as_peer_address)
+            .filter_map(
+                |listening_address| match (&receiver_address, listening_address) {
+                    (PeerAddress::Ip4(receiver), PeerAddress::Ip4(listener)) => {
+                        Some(PeerAddress::Ip4(PeerAddressIp4 {
+                            ip: receiver.ip,
+                            port: listener.port,
+                        }))
+                    }
+                    (PeerAddress::Ip6(receiver), PeerAddress::Ip6(listener)) => {
+                        Some(PeerAddress::Ip6(PeerAddressIp6 {
+                            ip: receiver.ip,
+                            port: listener.port,
+                        }))
+                    }
+                    _ => None,
+                },
+            )
+            .filter_map(|address| TransportAddress::from_peer_address(&address))
+            .collect::<Vec<_>>();
+
+        for address in discovered_own_addresses {
+            self.send_announced_address(peer_id, address).await?;
+        }
+
+        Ok(())
+    }
+
+    async fn send_announced_address(
+        &mut self,
+        peer_id: T::PeerId,
+        address: T::Address,
+    ) -> crate::Result<()> {
+        let peer_addresses = self.announced_addresses.entry(peer_id).or_default();
+        if !peer_addresses.contains(&address) {
+            self.peer_connectivity_handle
+                .send_request(
+                    peer_id,
+                    PeerManagerRequest::AnnounceAddrRequest(AnnounceAddrRequest {
+                        address: address.as_peer_address(),
+                    }),
+                )
+                .await?;
+            peer_addresses.insert(address);
+        }
+        Ok(())
+    }
+
     /// Handle connection established event
     ///
     /// The event is received from the networking backend and it's either a result of an incoming
     /// connection from a remote peer or a response to an outbound connection that was initiated
     /// by the node as result of the peer manager maintenance.
-    fn accept_connection(
+    async fn accept_connection(
         &mut self,
         address: T::Address,
         role: Role,
         info: PeerInfo<T::PeerId>,
-        _receiver_address: Option<PeerAddress>,
+        receiver_address: Option<PeerAddress>,
```

Comment:
Why is this an Option? Is it because of Inbound/Outbound? In that case, maybe we should put that in the Role enum. I'm not sure. But keeping variables that represent unexplainable states as Option is not really good. Another non-invasive option is, there could be a `ReceiverAddress` enum, which can have two possible values, Inbound(PeerAddress), and Outbound with no associated value. This way, the reader will immediately understand that this doesn't have a value because it's outgoing.

Reply:
No, it's in `Option` only because remote peers might decide not send it in handshake:

```
pub enum HandshakeMessage {
    Hello {
    ...
        /// Socket address of the remote peer as seen by this node (addr_you in bitcoin)
        receiver_address: Option<PeerAddress>,
    },
    HelloAck {
    ...
        /// Socket address of the remote peer as seen by this node (addr_you in bitcoin)
        receiver_address: Option<PeerAddress>,
    },
}
```

So there is no difference in inbound and outbound peers on a protocol level.

Reply:
So it's optional? Peers can choose not to send it? Weird.

Reply:
If someone is behind a proxy, sending the TCP address of the remote peer would be wrong (although using proxies is not currently supported).

Reply:
This actually reminds me that we should add support for SOCKS proxy. I hope our design doesn't make it so hard.

Reply:
OK, I will take a note.